weasyl/searchtag.py

```diff
@@ -217,3 +254,233 @@ def tag_history(submitid):
         .select_from(tu.join(pr, tu.c.userid == pr.c.userid))
         .where(tu.c.submitid == submitid)
         .order_by(tu.c.updated_at.desc()))
+
+
+def add_and_get_searchtags(tags):
+    """
+    Handles addition of, and getting existing, searchtags, abstracting the logic
+    for addition of such from editing functions. Serves to consolidate otherwise
+    duplicated code.
+
+    Parameters:
+        tags: A set of tags.
+
+    Returns:
+        query: The results of a SQL query which contains tagids and titles for
+                 tags which either currently exist, or were added as a result
+                 of this function.
+    """
+    # Get the tag titles/ids out of the searchtag table
+    query = d.engine.execute("""
+        SELECT tagid, title FROM searchtag WHERE title = ANY (%(title)s)
+    """, title=list(tags)).fetchall()
+
+    # Determine which (if any) of the valid tags are new; add them to the searchtag table if so.
+    newtags = list(tags - {x.title for x in query})
+    if newtags:
+        query.extend(
+            d.engine.execute(
+                "INSERT INTO searchtag (title) SELECT * FROM UNNEST (%(newtags)s) AS title RETURNING tagid, title",
+                newtags=newtags
+            ).fetchall())
+    return query
+
+
+def edit_user_searchtag_blacklist(userid, tags):
+    """
+    Edits the user searchtag blacklist, adding or removing tags as appropriate.
+
+    Parameters:
+        userid: The userid of the user submitting the request.
+        tags: A set() object of tags; must have been passed through ``parse_blacklist_tags()``
+                (occurs in the the controllers/settings.py controller)
+
+    Returns:
+        Nothing.
+    """
+    # Determine what, if any, tags exist before editing
```

Comment:
There’s no history here (unlike with submission tags); just delete and reinsert everything.